bob-backend/src/pearl/group.rs

```diff
@@ -302,32 +338,41 @@ impl Group {
         let holders = self.holders.read().await;
         let mut max_timestamp = None;
         let mut result = false;
+
         for (ind, &key) in keys.iter().enumerate() {
             for (_, Leaf { data: holder, .. }) in holders.iter_possible_childs_rev(&Key::from(key))
             {
-                match holder.exist(key).await.unwrap_or(ReadResult::NotFound) {
-                    ReadResult::Found(ts) => {
-                        let ts = ts.into();
-                        if ts > max_timestamp.unwrap_or(0) {
-                            max_timestamp = Some(ts);
-                            result = true;
+                if self.should_check_holder(holder, max_timestamp) {
+                    match holder.exist(key).await.unwrap_or(ReadResult::NotFound) {
+                        ReadResult::Found(ts) => {
+                            let ts = ts.into();
+                            if ts > max_timestamp.unwrap_or(0) {
+                                max_timestamp = Some(ts);
+                                result = true;
+                            }
                         }
-                    }
-                    ReadResult::Deleted(ts) => {
-                        let ts = ts.into();
-                        if ts > max_timestamp.unwrap_or(0) {
-                            max_timestamp = Some(ts);
-                            result = false;
+                        ReadResult::Deleted(ts) => {
+                            let ts = ts.into();
+                            if ts > max_timestamp.unwrap_or(0) {
+                                max_timestamp = Some(ts);
+                                result = false;
+                            }
                         }
+                        ReadResult::NotFound => continue,
                     }
-                    ReadResult::NotFound => continue,
                 }
             }
             exist[ind] = result;
         }
         Ok(exist)
     }
 
+    #[inline]
+    fn should_check_holder(&self, holder: &Holder, max_timestamp: Option<u64>) -> bool {
```

Comment:
Please, make this function static: don't pass `self` as parameter, pass `self.safe_timestamp_step` instead